### PR TITLE
Proposed specification update to support context sharing

### PR DESCRIPTION
**Pull request by rikoe**
This is a work-in-progress proposal to facilitate discussion for context sharing support in the FDC3 API.

Outstanding tasks:
- [x] Update context sharing description in specification
- [x] Add new methods to `src/api/interface.ts`
- [ ] Add new methods to `docs/api/DesktopAgent.md`
- [ ] Add usage examples
- [ ] Create separate PR for context channel support, which is now independent

### CODE REVIEW

File: docs/api/api-spec.md [hunk 1]
```diff
@@ -126,12 +126,78 @@ It is expected that App Directories will also curate listed apps and ensure that
...
+```ts
+const instrumentListener = desktopAgent.addTypedContextListener('fdc3.instrument', instrument => {
+    // handle instrument object
+});
+```
+
+If an application would like to retrieve the current state of the desktop at startup, they can ask for the current value for a particular type of object:
+
+```ts
+await instrument = desktopAgent.getCurrentValue('fdc3.instrument');
```
Comment (ColinEberhardt) [inline docs/api/api-spec.md]: `getCurrentValue` doesn't really tell you much about what this function does. I'd suggest `getContext` or `getContextValue` - the fact that it is current is implicit.

File: docs/api/api-spec.md [hunk 2]
```diff
@@ -126,12 +126,78 @@ It is expected that App Directories will also curate listed apps and ensure that
...
+```ts
+const listener = desktopAgent.addContextListener(context => {
+    // handle context object
+});
+```
+
+If an application declares a typed context listener, the callback will only be invoked for context object updates where the type matches:
+
+```ts
+const instrumentListener = desktopAgent.addTypedContextListener('fdc3.instrument', instrument => {
```
Comment (ColinEberhardt) [inline docs/api/api-spec.md]: Similar to @nkolba - I'd vote for overloading `addContextListener` to allow both typed and non-typed subscriptions.

File: docs/api/api-spec.md [hunk 3]
```diff
@@ -126,12 +126,78 @@ It is expected that App Directories will also curate listed apps and ensure that
...
+
+If an application declares a generic context listener, providing a callback, they will be notified whenever another application invokes the `broadcast` function for any object:
+
+```ts
+const listener = desktopAgent.addContextListener(context => {
+    // handle context object
+});
+```
+
+If an application declares a typed context listener, the callback will only be invoked for context object updates where the type matches:
```
Comment (ColinEberhardt) [inline docs/api/api-spec.md]: > where the type matches

Can you provide more clarity on this? is a match string equality? Or something more complex (e.g. matching a pattern or hierarchy?)

Comment (rikoe) [reply docs/api/api-spec.md]: The intention was just string equality for now. I think if a particular implementation would like to perform more complicated matching, they could, but the spec should not mandate it.

But in any case, will make this more explicit.

**Pull request closed without merging**